File: Spring/Chapter4/java_demo/demo/src/main/java/com/cglib/CGLibProxy.java
package com.cglib;

import net.sf.cglib.proxy.Enhancer;
import net.sf.cglib.proxy.MethodInterceptor;
import net.sf.cglib.proxy.MethodProxy;

import java.lang.reflect.Method;

public class CGLibProxy implements MethodInterceptor {
    public Object createProxy(Object target)
    {
        Enhancer enhancer = new Enhancer();
        enhancer.setSuperclass(target.getClass());
        enhancer.setCallback(this);
        return enhancer.create();
    }

    @Override
    public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy) throws Throwable {
        CGLibAspect aspect = new CGLibAspect();
        aspect.check();
        aspect.except();
        Object obj = methodProxy.invokeSuper(o,objects);
        aspect.log();
        aspect.monitor();
        return obj;
    }
}

File: Spring/Chapter4/kotlin_demo/demo/src/main/kotlin/cglib/CGLibProxy.kt
package cglib

import net.sf.cglib.proxy.Enhancer
import net.sf.cglib.proxy.MethodInterceptor
import net.sf.cglib.proxy.MethodProxy
import java.lang.reflect.Method

class CGLibProxy : MethodInterceptor{

    fun createProxy(target: Any):Any
    {
        val enhancer = Enhancer()
        enhancer.setSuperclass(target.javaClass)
        enhancer.setCallback(this)
        return enhancer.create()
    }

    override fun intercept(obj: Any?, method: Method?, args: Array<out Any>?, proxy: MethodProxy?): Any?{
        val aspect = CGLibAspect()
        aspect.check()
        aspect.expect()
        val o = proxy?.invokeSuper(obj,args)
        aspect.log()
        aspect.monitor()
        return o
    }
}